File: array/0078.-Subsets.go
/**
  题目:
    给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。
    说明：解集不能包含重复的子集。
    示例:

  输入: nums = [1,2,3]
  输出:
  [
    [3],
    [1],
    [2],
    [1,2,3],
    [1,3],
    [2,3],
    [1,2],
    []
  思路:
    定义选择列表 nums 选择路径path 结果集 res 以及 进行去重复的idx
  效果:
    执行用时：0 ms, 在所有 Go 提交中击败了100.00%的用户
    内存消耗：2.3 MB, 在所有 Go 提交中击败了33.59%的用户
*/

func subsets(nums []int) [][]int {
    if len(nums) == 0{
        return  nil
    }
    res := make([][]int, 0)
    helper(nums,&res, []int{},0)
    return res
}
func helper(nums []int,res *[][]int, path []int, idx int){
    *res = append(*res, append([]int{},path...))
    for i := idx; i < len(nums);i++{
        path = append(path, nums[i])
        helper(nums, res, path, i+1)
        path = path[:len(path)-1]
    }
    return
}

File: array/0078.-Subsets.cpp
/**
  题目:
      给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。

    说明：解集不能包含重复的子集。

    示例:

        输入: nums = [1,2,3]
        输出:
          [
            [3],
            [1],
            [2],
            [1,2,3],
            [1,3],
            [ 2,3],
            [1,2],
            []
         ]

 
 链接：
   https://leetcode-cn.com/problems/subsets 
思路:
     1.典型的dfs+递归
    注意：
    可以看看排列和组合
 
       
  效果:
    执行用时：8 ms, 在所有 C++ 提交中击败了9.97%的用户
    内存消耗：7 MB, 在所有 C++ 提交中击败了55.45%的用户
*/ 


class Solution {
public:
vector<vector<int>> res;
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> path;
        subsets(nums,0,nums.size(),path);
        return res;
    }
    void subsets(vector<int>&nums,int index,int size,vector<int>&path){
        res.push_back(path);
        for(int i=index;i<size;i++){
            path.push_back(nums[i]);
            subsets(nums,i+1,size,path);
            path.pop_back();
        }
    }
};